### PR TITLE
MAID-2454 create a votes iterator for Block to transform proofs into votes

### PR DESCRIPTION
**Pull request by maqi**

### CODE REVIEW

File: src/vote.rs [hunk 1]
```diff
@@ -44,6 +44,11 @@ impl<T: Serialize + Clone> Vote<T> {
         })
     }
 
+    /// Compose a Vote.
+    pub fn compose(payload: T, signature: Signature) -> Self {
```
Comment (madadam) [inline src/vote.rs]: This function allows you to create a `Vote` where the signature doesn't match the payload. I'd consider marking it `unsafe`, but not entirely sure about that. At the very least, do not expose it outside of the crate, so make it `pub(crate)` or something.

File: src/block.rs [hunk 2]
```diff
@@ -156,6 +156,17 @@ impl<T: Serialize + Clone> Block<T> {
         }
     }
 
+    /// Create an iterator over all proofs and transform into votes.
+    pub fn votes_iter<'a>(&'a self) -> Box<Iterator<Item = (&PublicInfo, Vote<T>)> + 'a> {
+        let payload = self.payload.clone();
+        Box::new(self.proofs.iter().map(move |proof| {
+            (
+                proof.node_info(),
+                Vote::compose(payload.clone(), *proof.sig()),
```
Comment (Fraser999) [inline src/block.rs]: Can't we just use `self.payload.clone()` here rather than taking another copy outside the loop?

File: src/vote.rs [hunk 3]
```diff
@@ -44,6 +44,11 @@ impl<T: Serialize + Clone> Vote<T> {
         })
     }
 
+    /// Compose a Vote.
+    pub(crate) fn compose(payload: T, signature: Signature) -> Self {
```
Comment (Fraser999) [inline src/vote.rs]: We've spoken about making a module for the data chain stuff before.  If we did that now, we could reduce the visibility of this method to just that module.

Comment (madadam) [reply src/vote.rs]: That would be even better.

File: src/block.rs (hunk 2 above)
Comment (maqi) [reply src/block.rs]: will that `move` the whole `self` ?

Comment (Fraser999) [reply src/block.rs]: No, I don't think so - we only have a reference to `self` there.

**Pull request merged**